include/net/nrf_cloud_pgps.h

```diff
@@ -315,12 +315,22 @@ void nrf_cloud_pgps_request_reset(void);
  * @param buf Pointer to data received from nRF Cloud.
  * @param buf_len Buffer size of data to be processed.
  *
- * @retval 0 A-GPS data successfully processed.
+ * @retval 0 P-GPS data successfully processed.
  * @retval -EFAULT An nRF Cloud P-GPS error code was processed.
  * @return A negative value indicates an error.
  */
 int nrf_cloud_pgps_process(const char *buf, size_t buf_len);
 
+/** @brief Processes binary P-GPS data using URL received from nRF Cloud.
+ *
+ * @param file_location Pointer to structure holding the hostname and filename.
+ *
+ * @retval 0 P-GPS data successfully processed.
+ * @retval -EFAULT An nRF Cloud P-GPS error code was processed.
+ * @return A negative value indicates an error.
+ */
+int nrf_cloud_pgps_update(struct nrf_cloud_pgps_result *file_location);
```

Comment:
I wrote a comment for this already yesterday, but forgot to submit the review, sorry.

I think memory handling with `struct nrf_cloud_pgps_result` could use some changes. At first I didn't notice that user has to provide buffers for `host` and `path`, but just initialized `struct nrf_cloud_pgps_result` to zero and gave it to `nrf_cloud_coap_pgps_url_get()`. Obviously, this didn't work and yes, `nrf_cloud_pgps.h` says, that these are user-provided buffers. However, it does not say anything about the lifetime of the buffers. My first idea was to use stack, but then started looking into how the memory is actually used by the implementation.

It seems that pointers to those buffers are passed directly to Download client, which stores and uses them until the HTTP download is finished. The simplest way is to do as you have here, i.e. use static buffers, but I don't think this is very nice.

IMHO, the library should hide this and remove the need for the user to allocate (and possibly free) memory for the host and path.

Reply:
Isn't this the same with `nrf_cloud_rest_pgps_data_get()`?

Reply:
In the REST case, the buffer is pointed to by the rest context, and it is up to the caller to provide it.

Reply:
You mean `rx_buf`? Yes, it works in the same way. It's just that chunks of RAM are reserved for something that are used only occasionally, maybe as rarely as once a week.

With A-GPS it's not that bad, because it would be easy to allocate the memory dynamically and know when it can be freed.

With P-GPS that's not so obvious, and we don't even have an example for it. I'm not sure how the application can determine that the download has ended and the buffers (`rx_buf`, `host `and `path`) can be freed. `PGPS_EVT_READY`? Is that event sent also if the download fails? The library knows when the buffer is no longer needed. It would also make the API usage simpler, if the application would not have to worry about the memory handling.

In modem_shell, we have all three buffers twice, because A-GPS and P-GPS support is present in both Location library and the modem_shell "gnss" command. That's 4480 bytes of RAM reserved only for this. Of course modem_shell is just a test application, but just trying to make a point that we do have quite a limited amount of RAM available.

Reply:
What I would like to do for this PR is to update the documentation to explain what the lifetime of the user-provided memory needs to be, and when it is safe to free the memory, if the user-provided memory was dynamically allocated.

The task of refactoring both the REST and CoAP handling of P-GPS buffers is out of scope of this PR.